File: javascript/balancing_parentheses.py
def balancing_parentheses(string):
    openings = 0
    missing = 0

    for e in string:
        if e == "(":
            openings += 1
        else:
            if openings > 0:
                openings -= 1
            else:
                missing += 1

    return openings + missing

print("Expecting: 0")
print(balancing_parentheses("(()())"))

print("")

print("Expecting: 2")
print(balancing_parentheses("()))"))

print("")

print("Expecting: 1")
print(balancing_parentheses(")"))

File: javascript/balancing_parentheses.js
function balancingParentheses(string) {
  let openings = 0;
  let missing = 0;

  for (let i = 0; i < string.length; ++i) {
    if (string[i] === "(") {
      ++openings;
      continue;
    }

    if (openings > 0) {
      --openings;
    } else {
      ++missing;
    }
  }
  return openings + missing;
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 0");
  console.log(balancingParentheses("(()())"));

  console.log("");

  console.log("Expecting: 2");
  console.log(balancingParentheses("()))"));

  console.log("");

  console.log("Expecting: 1");
  console.log(balancingParentheses(")"));
}

module.exports = balancingParentheses;

// Please add your pseudocode to this file
// Declare constant "count" equal to 0
// Check the first character in the string: if it's an open, add 1 to count; if close, subtract 1 from count
// recurse using the string, minus its first letter
// return the absolute value of count
// And a written explanation of your solution
